lib/clients/AliClient.py

```diff
@@ -69,4 +154,397 @@ def _download_from_blobstore(self, blob_to_download_name, blob_download_target_p
+            self.logger.error(message)
+            if snapshot or snapshot_creation_operation:
+                self.delete_snapshot(snapshot_id)
+                snapshot = None
+            raise Exception(message)
+
+        return snapshot
+
+    def _get_snapshot_list(self, snapshot_id):
+        region_id = self.__aliCredentials['region_name']
+        get_snapshot_req_params = {
+            'PageSize' : 10,
+            'RegionId' : region_id,
+            'SnapshotIds' : [snapshot_id]
+        }
+        get_snapshot_request = self._get_common_request('DescribeSnapshots', get_snapshot_req_params)
+        snapshot_details = self.compute_client.do_action_with_exception(get_snapshot_request)
+        snapshot_details_json = json.loads(snapshot_details.decode('utf-8'))
+        return snapshot_details_json['Snapshots']['Snapshot']
+    
+    def _get_snapshot_with_exception(self, snapshot_id):
+        snapshot_list = self._get_snapshot_list(snapshot_id)
+        if len(snapshot_list) > 1:
+            message = 'More than 1 snapshot found with id {}'.format(
+            snapshot_id)
+            raise Exception(message)
+        if len(snapshot_list) == 0:
+            message = 'Snapshot with id {} is not found'.format(
+            snapshot_id)
+            raise Exception(message)
+        snapshot = snapshot_list[0]
+        return Snapshot(snapshot['SnapshotId'], snapshot['SourceDiskSize'], snapshot['CreationTime'], snapshot['Status'])
+
+    def get_snapshot(self, snapshot_id):
+        try:
+            return self._get_snapshot_with_exception(snapshot_id)
+        except Exception as error:
+            self.logger.error('[ALI] ERROR: Error in getting snapshot {}.\n{}'.format(
+                snapshot_id, error))
+            return None
+    
+    def _is_snapshot_ready(self, snapshot_id):
+        """Gets the snapshot state.
+        https://www.alibabacloud.com/help/doc-detail/25641.htm?spm=a2c63.p38356.a3.5.c880458dkimfOs#SnapshotType
+        Status can be "progressing" "accomplished" or "failed"
+        """
+        # Try until snapshot state is either "accomplished" or "failed"
+        # Returns True if multiple snapshots with same id are found
+        # Retries for failures also
+        try:
+            snapshot_list = self._get_snapshot_list(snapshot_id)
+            if len(snapshot_list) > 1 or (len(snapshot_list) == 1 and snapshot_list[0]['Status'] in ('accomplished','failed')):
+                return True
+            return False
+        except Exception as error:
+            self.logger.error(
+                '[ALI] ERROR: Unable to get snapshot details for snapshot id {}.\n{}'.format(snapshot_id, error))
+            return False
+
+    def _copy_snapshot(self, snapshot_id):
+        return self.get_snapshot(snapshot_id)
+
+    def _delete_snapshot(self, snapshot_id):
+        log_prefix = '[SNAPSHOT] [DELETE]'
+        try:
+            snapshot_deletion_req_params = {
```

Comment:
A/c to [documentation](https://www.alibabacloud.com/help/doc-detail/25525.htm?spm=a2c63.p38356.b99.470.2ecd123fj8j9jB)
> You cannot delete the snapshot if it has been used to create a custom image (CreateImage). However, you can delete the appropriate custom image (DeleteImage) before you try the DeleteSnapshot action again

However we can pass flag {'Force': True} to forcefully delete snapshot, but disk created from the snapshot cannot be reinitialized

https://www.alibabacloud.com/help/doc-detail/25449.htm#h2-url-2
> If the data disk was created from a snapshot, then the data disk is recovered to the point in time at which the snapshot was generated. You do not have to mount the partitions again, but all the data generated after the disk creation is lost.